pkg/kubelet/kubelet.go

```diff
@@ -1776,6 +1779,50 @@ func (kl *Kubelet) pastActiveDeadline(pod *api.Pod) bool {
 	return false
 }
 
+// Get pods which should be resynchronized. Move the logic here so as to make it testable and make the main sync loop
+// simple and clear.
+// Currently, the following pod should be resynchronized:
+//   * pod whose work is ready.
+//   * pod past the active deadline.
+func (kl *Kubelet) getPodsToSync() []*api.Pod {
+	// Get all pods past the active deadline
+	allPods := kl.podManager.GetPods()
+	var podsToFail []*api.Pod
+	for _, pod := range allPods {
+		if !kl.pastActiveDeadline(pod) {
+			continue
+		}
+		podsToFail = append(podsToFail, pod)
+	}
+
+	// Get all pods whose work are ready
+	podUIDs := kl.workQueue.GetWork()
+	var podsToWork []*api.Pod
+	for _, uid := range podUIDs {
+		if pod, ok := kl.podManager.GetPodByUID(uid); ok {
+			podsToWork = append(podsToWork, pod)
+		}
+	}
+
+	// Merge the pod lists
```

Comment:
The merging code is abit hard to expand if we add more criteria along the way. My suggestion would be:
 1. Get all the pods at the beginning of this function.
 2. Get the podUIDs from the work queue, and create a set of uids.
 3. Iterate through every pod from (1), and check whether they matches any of the following criteria; append if it's true
    - pastActiveDeadline == true
    - uid in the set from (2)
4. Return the list from (3)